wpilib/src/robot.rs

```diff
@@ -0,0 +1,155 @@
+        (_, RobotState::Disabled) => robot.disabled_init(),
+
+        (Some(RobotState::Autonomous), RobotState::Autonomous) => (),
+        (_, RobotState::Autonomous) => robot.autonomous_init(),
+
+        (Some(RobotState::Teleop), RobotState::Teleop) => (),
+        (_, RobotState::Teleop) => robot.teleop_init(),
+
+        (Some(RobotState::Test), RobotState::Test) => (),
+        (_, RobotState::Test) => robot.test_init(),
+
+        // XXX: This really should look the same as disabled
+        (_, RobotState::EStop) => (),
+    }
+
+    // Call the appropriate periodic function
+    match cur_mode {
+        RobotState::Autonomous => {
+            unsafe { HAL_ObserveUserProgramAutonomous() }
+            robot.autonomous_periodic()
+        }
+        RobotState::Teleop => {
+            unsafe { HAL_ObserveUserProgramTeleop() }
+            robot.teleop_periodic()
+        }
+        RobotState::Test => {
+            unsafe { HAL_ObserveUserProgramTest() }
+            robot.test_periodic()
+        }
+        _ => {
+            unsafe { HAL_ObserveUserProgramDisabled() }
+            robot.disabled_periodic()
+        }
+    }
+
+    robot.robot_periodic()
+}
+
+/// Start the main robot loop for an IterativeRobot.
+/// The periodic methods are called each time a new packet
+/// received from the driver station.
+///
+/// It is recommended to use `start_timed` instead,
+/// which guarantees a more regular period of execution.
+pub fn start_iterative<T: IterativeRobot>(robot: &mut T, ds: &DriverStation) {
+    let mut last_mode: Option<RobotState> = None;
+
+    println!("\n********** Robot program starting **********\n");
+    unsafe { HAL_ObserveUserProgramStarting() }
+
+    loop {
+        ds.wait_for_data();
+
+        let cur_mode = ds.robot_state();
+        loop_func(robot, last_mode, cur_mode);
+        last_mode = Some(cur_mode);
+    }
+}
+
+/// Start the main robot loop for an IterativeRobot.
+/// The periodic methods are called every 20 milliseconds.
+///
+/// If you wish to have your main loop run at a different rate,
+/// use `start_timed_with_period`.
+pub fn start_timed<T: IterativeRobot>(robot: &mut T, ds: &DriverStation) {
+    start_timed_with_period(robot, ds, time::Duration::from_millis(20))
+}
+
+/// Start the main robot loop for an IterativeRobot.
+/// The periodic methods are called on a regular interval specified by `period`.
+pub fn start_timed_with_period<T: IterativeRobot>(
+    robot: &mut T,
+    ds: &DriverStation,
+    period: time::Duration,
+) {
+    let mut last_mode: Option<RobotState> = None;
+    let notifier = Alarm::new().expect("Failed to initialize FPGA notifier");
+    let period = period.as_micros() as u64;
+
+    println!("\n********** Robot program starting **********\n");
+    unsafe { HAL_ObserveUserProgramStarting() }
+
+    let mut expiration_time =
+        RobotBase::fpga_time().expect("Failed to read current FPGA time") + period;
+    let _ = notifier.update(expiration_time);
```

Comment:
Why is this unused `HalResult<()>` being saved?

Reply:
`let _ = ` is the idiom to throw away `must_use` values that are unused. You can't actually use the variable name `_`, iiuc.